### PR TITLE
initial workaound for IPv6 workload support

### PR DESCRIPTION
**Pull request by bonnyr**
### Proposed changes
IPv6 support in k8s is hitting mainstream and at some point IPv6 will become first class citizen in k8s deployments. 
Ingress Controller code currently assumes that all workload addresses are IPv4, leading to failures using IC in IPv6 enabled
clusters (or pure IPv6 clusters - see robin.io).

Theres a bug addressing this : https://github.com/nginxinc/kubernetes-ingress/issues/991

Turns out the fix is rather simple (at least when validated against a pure IPv6 cluster).

The use cases that this fix can address include:
1. IPv6 clusters where workloads (Service Endpoints mapped to upstreams)  are assigned IPv6 addresses. 
2. Dual Stack clusters where workloads are assigned IPv4/IPv6 addresses
   Note that the change below (and the underlying assumption of IC) wrt an endpoint having a single address will need to be tested
   

####Note
At this stage testing has been performed manually on an IPv6 cluster with a simple nginx deployment serving as the
workload. The test environment has shown that:
1. the initial nginx configuration generated includes the right format IP address for the upstreams 
2. scaling the deployment up and using the Plus API to add a new entry resulted in no error
3. scaling the deployment down and using the Plus API to remove the deleted entry resulted in no error

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works - TBD
- [x] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation - N/A ?
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by centromere** [comment]
Thank you!

**Comment by centromere** [comment]
@bonnyr The following changes were also needed in my cluster.

https://github.com/centromere/kubernetes-ingress/commit/ee584a62f00221a935114118ddc2cafe0ae1752c

**Comment by bonnyr** [comment]
> @bonnyr The following changes were also needed in my cluster.
> 
> [centromere@ee584a6](https://github.com/centromere/kubernetes-ingress/commit/ee584a62f00221a935114118ddc2cafe0ae1752c)

Thanks for the heads up. This was originally ignored but I have added this to facilitate unit tests

File: internal/configs/ingress.go [hunk 1]
```diff
@@ -456,10 +456,10 @@ func createUpstream(ingEx *IngressEx, name string, backend *networking.IngressBa
 		}
 
 		for _, endp := range endps {
-			addressport := strings.Split(endp, ":")
+			pos := strings.LastIndex(endp, ":")
```
Comment (Frankkkkk) [inline internal/configs/ingress.go:459]: Why not use `net.SplitHostPort`

File: internal/k8s/controller.go [hunk 2]
```diff
@@ -2991,6 +2991,13 @@ func getEndpointsBySubselectedPods(targetPort int32, pods []*api_v1.Pod, svcEps
 	return endps
 }
 
+func ipv6SafeAddrPort(addr string, port int32) string {
```
Comment (Frankkkkk) [inline internal/k8s/controller.go:2994]: same here. Woudn't using `net.JoinHostPort` be cleaner ?

File: internal/configs/ingress.go (hunk 1 above)
Comment (bonnyr) [reply internal/configs/ingress.go:459]: Absolutely - good catch

File: internal/k8s/controller.go (hunk 2 above)
Comment (bonnyr) [reply internal/k8s/controller.go:2994]: ditto. Thanks

File: build/Dockerfile [hunk 3]
```diff
@@ -48,7 +48,7 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
 	&& apt-get install -y bsdutils mount util-linux libuuid1 libmount1 libblkid1 libsmartcols1 \
 	&& curl -fsSL https://cs.nginx.com/static/keys/nginx_signing.key | gpg --dearmor > /etc/apt/trusted.gpg.d/nginx_signing.gpg \
 	&& curl -fsSL -o /etc/apt/apt.conf.d/90pkgs-nginx https://cs.nginx.com/static/files/90pkgs-nginx \
-	&& DEBIAN_VERSION=$(awk -F '=' '/^VERSION_CODENAME=/ {print $2}' /etc/os-release) \
+	&& DEBIAN_VERSION=bullseye \
```
Comment (centromere) [inline build/Dockerfile:51]: What is the purpose of this change?

**Comment by centromere** [comment]
@bonnyr I needed [these changes](https://github.com/centromere/kubernetes-ingress/commit/0acf54bc161396eb9f9c9120de90174b1ebdee58) also.

**Comment by centromere** [comment]
Hi all. Is there any more development required for this to be accepted/merged?

**Comment by jjngx** [comment]
Hi @centromere, I started working on this PR. We need to add unit and automated acceptance tests for this feature. This is WIP.

**Comment by brianehlert** [comment]
@gattytto When this has met our quality bar and merged it will be consumable using our 'edge' release prior to waiting for the next milestone release.

**Comment by lucacome** [comment]
Closing this as #2576 was merged
